# Comparison of Lightweight Methods for Clustering on SNP Data ($k$-means vs. hierarchical)

## Overview

This project investigates the performance of traditional $k$-means clustering and hierarchical clustering on SNP data to capture population structure in the 1000 Genomes dataset. The dataset contains samples from numerous populations and subpopulations whose structure is captured in the corresponding SNP data. We compare the performances of traditional $k$-means clustering and hierarchical clustering in regards to capturing this structure with principal components generated from the SNP data via `plink`.

## Group Members

* Isaac Thomas | Grad Student, CSE

## Repository Contents

The repository is structured as follows:
* `data` - contains data for PCA. To save time, we only include SNP data from chromosome 19.
* `notebooks` - contains the notebook used for clustering
* `out` - contains analysis output report
* `plink` - contains the `plink` binary
* `report` - contains the written report for this analysis
* `scripts` - contains scripts for cleaning directories and running the pipeline 
* `tmp` - contains intermediate data generated by `plink`


## Usage

We strongly recommend that you run this analysis in a Linux environment. Command line tools like `realpath` or `readlink` are used to avoid relative pathing in the pipeline; these utilities may not be available in compatible versions or at all on other systems.

It is first recommended to run `bash scripts/clean.sh` for sanity.

The script `scripts/pipe.sh` runs the entire pipeline:
* PCA with `plink`
* clustering, hyperparameter tuning and visualization
* notebook export to markdown

you can specify multiple optional parameters:
 * `-p <num_principal_components>` 
 * `-k <max_number_clusters>`
 * `-c <clustering_iterations>`
 * `-m <minor_allele_frequency>`
 * `-s <random_seed>`
 * `-v yes` (explained below)

`bash scripts/pipe.sh -p <num_principal_components> -k <max_number_of_clusters> -c <clustering_iterations> -m <minor_allele_frequency> -s <random_seed> [-v yes]`

## Dependencies

This analysis depends on:
* `plink` for PCA and SNP variance weighting
* `numpy` and `pandas` for data manipulation 
* `scikit-learn` for clustering
* `matplotlib` for plotting 
* `papermill` and `nbconvert` for notebook execution/export

`plink` is included in the repository. Additionally, the pipe script can optionally create and activate an environment with the remaining dependencies. Just add `-v yes` to the command options. As an example:

`bash scripts/pipe.sh -p 4 -k 12 -m 0.05 -v yes`

will create a `conda` environment and activate it for the duration of the analysis.

## Downloading 1000 Genomes Dataset

If you would like to try this analysis on SNPs from other chromosomes, you can do the following:
```
cd data
mkdir -p thousand_genomes_data
#download the entire data folder
wget -r -np -m --no-check-certificate -P $(pwd)/thousand_genomes_data ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/data_collections/1000_gen
omes_project/release/20181203_biallelic_SNV/
```
you will then need to LD-prune the data (we are already using pruned data) and move all of the resulting files back into the `data` directory. This is a time-consuming process and is out of the scope of this analysis, but you can refer [here](https://www.cog-genomics.org/plink/1.9/ld) to do so with `plink`.

## Results

We performed $k$-means clustering on the SNP data to determine how well it could capture population and subpopulation level structures. It came close to capturing the true population structure (silhouette score of $\sim 0.75$ with an optimal 5 clusters), while tuning to cluster at the subpopulation level yields poor precision (silhouette score of $\sim 0.26$ with an "optimal" 20 clusters). In this case, there are likely many mislabellings between subpopulations belonging to the same population.

The best instances of hierarchical clustering (using Ward linkage and average linkage) achieved a similar result when tuned by distance threshold instead of desired number of clusters (sil. scores of ~0.75, optimal $k$ was also 5). Both of these methods also did not discern subpopulation structure well. 

As visualized in the analysis, the common shortcoming of these methods is likely explained by the unimodal distribution of pairwise distances <u>within</u> populations (no big difference between subpopulation-level intracluster and intercluster distance).

## Limitations and Future Research

This project has the following limitations:
* use of data from only one chromosome
* use of genetically/biologically uninformed distance functions
* use of naive hyperparameter tuning methods 

Data from more chromosomes could yield better separation via subpopulation-specific SNPs, and genetically informed distance functions could take advantage of this richer representation. Additionally, future work should be done on a way to combine the results of hierarchical clustering at multiple distance thresholds to capture subcluster structure more effectively.
