# Comparison of Lightweight Methods for Clustering on SNP Data ($k$-means vs. hierarchical)

## Overview

This project investigates the performance of traditional $k$-means clustering and hierarchical clustering on SNP data to capture population structure in the 1000 Genomes dataset. The dataset contains samples from numerous populations and subpopulations whose structure is captured in the corresponding SNP data. We compare the performances of traditional $k$-means clustering and hierarchical clustering in regards to capturing this structure with principal components generated from the SNP data via `plink`.

## Group Members

* Isaac Thomas | Grad Student, CSE

## Repository Contents

The repository is structured as follows:
* `data` - contains data for PCA. To save time, we only include SNP data from chromosome 19.
* `notebooks` - contains the notebook used for clustering
* `out` - contains analysis output report
* `plink` - contains the `plink` binary
* `report` - contains the written paper for this analysis
* `scripts` - contains scripts for cleaning directories and running the pipeline 
* `tmp` - contains intermediate data generated by `plink`


## Usage

We strongly recommend that you run this analysis in a Linux environment. Command line tools like `realpath` or `readlink` are used to avoid relative pathing in the pipeline; these utilities may not be available in compatible versions or at all on other systems.

It is first recommended to run `bash scripts/clean.sh` for sanity.

The script `scripts/pipe.sh` runs the entire pipeline:
* PCA with `plink`
* clustering, hyperparameter tuning and visualization
* notebook export to markdown

you can specify multiple optional parameters:
 * `-p <num_principal_components>` 
 * `-k <max_num_clusters>`
 * `-m <min_minor_allele_frequency>`

`bash scripts/pipe.sh -p <num_principal_components> -k <max_number_of_clusters> -m <minor_allele_frequency> [-v yes]`

## Dependencies

This analysis depends on:
* `plink` for PCA and SNP variance weighting
* `numpy` and `pandas` for data manipulation 
* `scikit-learn` for clustering
* `matplotlib` for plotting 
* `papermill` and `nbconvert` for notebook execution/export

`plink` is included in the repository. Additionally, the pipe script can optionally create and activate an environment with the remaining dependencies. Just add `-v yes` to the command options. As an example:

`bash scripts/pipe.sh -p 4 -k 12 -m 0.05 -v yes`

will create a `conda` environment and activate it for the duration of the analysis.

## Remaining Tasks
* Compare with hierarchical clustering
* Explore variance weights of SNPs from PCA output
* Add details about downloading 1000 Genomes dataset to readme